add expire and inactive options to linux user resource

Stromweld opened the pull request:
Signed-off-by: Corey Hemminger <hemminger@hotmail.com>

adds expiredate and inactive options to linux user resource

## Description
This allows additional options to be passed to set an expiration date on a created user as well as adjust or disable the inactive  days counter that disables an account after the password expires. The latter is common config in CIS benchmark to be set to 30 days over default of -1 which disables the feature.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

Stromweld commented:
@tas50 I got all current tests working, but not sure where to create new tests and verify new functionality. Would you be able to help with this last piece in anyway? Here's is the link on the  shadow ruby gem usage. https://github.com/apalmblad/ruby-shadow.

lamont-granquist commented on lib/chef/provider/user.rb:
```diff
@@ -66,13 +66,18 @@ def load_current_resource
+              expire_date = DateTime.strptime(shadow_info.sp_expire&.to_s, "%d")
+              current_resource.expire_date(expire_date.strftime("%Y-%m-%d"))
+            end
+            if new_resource.password && current_resource.password == "x"
```
I think you can avoid the new_resource.inactive/expire_dates here and rely on `should_set?` doing the nil check against the new_resource.  The current_resource should be the whole current state (although checking `== "x"` against the password really is current state of which password field is being used so that is correct).  The existing check around `new_resource.password` is mostly likely redundant and wrong I would think -- or should definitely also be moved later with a `should_set?` check around actual password setting which might be missing.

Stromweld replied:
To be honest I don't fully understand how this all works, just trying to repeat existing patterns in code for the 2 additional fields. The new_resource.password check was there prior so I didn't want to break anything by making extra changes that I'm not sure why it was there to begin with. Biggest issue I have is not having the ability to properly test these changes to ensure they work. Or atleast lack of knowledge on how to test. I believe I'm 90% of the way there. Any help you can provide to get this over the finish line would be greatly appreciated. The biggest feature we need is the inactive setting for CIS audit compliance. We have a workaround for now but figured it'd be good to get into the core resource and would be useful for others.

Stromweld commented:
@lamont-granquist I removed the if statements with the new_resources usage as suggested. I added them to get passed the pipeline errors but a more proper fix is probably better. Just not entirely sure how to fix.

lamont-granquist commented:
I think the problem is that `Shadow::Passwd.getspnam` returns nil when the user doesn't exist, so you need to wrap the rest of it with `unless shadow_info.nil? .... end` probably.  (You could try to `&.` protect things, but that looks like it probably gets messy with the expire_date).

Stromweld commented:
@lamont-granquist have you seen these kind of errors before?

Stromweld commented:
@lamont-granquist @tas50 I'm not sure how busy you guys are, but wondering if I could implore you guys for a little help getting this over the threshold?

lamont-granquist commented:
it looks like `shadow_info.sp_expire` is an integer and not a string, that might suggest to me that the ruby shadow library already converts it to epoch time and that the strptime is unnecessary?  just kind of guessing though based on the exception.  i don't know that i've ever needed to use strptime, and definitely don't know what ruby-shadow returns.

Stromweld commented:
@lamont-granquist you are correct, `shadow_info.sp_expire` returns an integer for days since epoch till account will expire. since expire date option is string in yyyy-mm-dd format I was trying to use strptime to convert the integer into a string date and then strftime in the next line to convert it to format similar to the users input for expire date so the comparison can evaluate it properly to verify if a change is needed.

Stromweld commented:
Now it looks like it's working and passing all tests except the one ssh_command spec test.

Stromweld commented:
@lamont-granquist This is passing all checks now. The 1 failure test looks to be from an internet blip while apt was updating packages.

tas50 commented on lib/chef/resource/user.rb:
```diff
@@ -50,6 +50,14 @@ class User < Chef::Resource
 
+      property :expire_date, [ String, NilClass ],
+        description: "(Linux) The date on which the user account will be disabled. The date is specified in the format YYYY-MM-DD.",
+        desired_state: false
```
These need `introduced: "17.6"` as well

lamont-granquist commented:
The places to look for testing are:

spec/unit/resource/user_spec.rb
spec/unit/provider/user_spec.rb
spec/unit/provider/user/linux_spec.rb

lamont-granquist commented:
I'm just leaving a comment here so this appears towards the top of the recently updated list so we look at it next week when phil is back.

Stromweld commented:
Thanks. If you send me a link I can also try to attend.

jaymzh commented on spec/unit/provider/user/linux_spec.rb:
```diff
@@ -34,6 +34,8 @@
     "shell" => "-s",
     "password" => "-p",
+    "expire_date" => "-e",
+    "inactive" => "-f",
```
Let's add unit tests below for these two new options, you should be able to copy the others.

Stromweld replied:
The current tests are only checking manage_home behavior. I'm not very familiar with writing these kinds of tests.

jaymzh replied:
See lines 61-73 - just tests to ensure that the right CLI options are set by default and change when people set the right resource propreties.

Stromweld replied:
unit tests added

jaymzh reviewed (changes requested):
This looks good, sorry for the massive delay. Let's rebase and then add the tests I recommend and update the "version added in" and we'll get this merged.

jaymzh commented on spec/unit/provider/user/linux_spec.rb:
```diff
@@ -34,15 +34,18 @@
 
   include_examples "a useradd-based user provider", supported_useradd_options
 
+  before(:each) do
+    @new_resource = Chef::Resource::User::LinuxUser.new("adam", @run_context)
+    @current_resource = Chef::Resource::User::LinuxUser.new("adam", @run_context)
+  end
```
the `before` needs to be inside of a `describe` block, so just copy this into each of the 3 describes.

Stromweld replied:
fixed

jaymzh commented on spec/unit/provider/user/linux_spec.rb:
```diff
@@ -70,4 +72,54 @@
     end
   end
+
+  describe "expire_date behavior" do
```
Sorry, these options are in `universal_options`, not `useradd_options`, my bad. This diff makes your tests pass for me:

```ruby
[phil@rider (Stromweld-master) RVM(2.7.1) chef]$ git diff
diff --git a/spec/unit/provider/user/linux_spec.rb b/spec/unit/provider/user/linux_spec.rb
index 8c2a7c52cd..61d7c43c6a 100644
--- a/spec/unit/provider/user/linux_spec.rb
+++ b/spec/unit/provider/user/linux_spec.rb
@@ -84,17 +84,17 @@ describe Chef::Provider::User::Linux do
     end
 
     it "by default expire_date is nil and we use ''" do
-      expect( provider.useradd_options ).to eql([""])
+      expect( provider.universal_options ).to eq([])
     end
 
     it "setting expire_date to nil includes ''" do
       @new_resource.expire_date nil
-      expect( provider.useradd_options ).to eql([""])
+      expect( provider.universal_options ).to eq([])
     end
 
     it "setting expire_date to 1982-04-16 includes -e" do
       @new_resource.expire_date "1982-04-16"
-      expect( provider.useradd_options ).to eql(["-e"])
+      expect( provider.universal_options ).to eq(["-e", "1982-04-16"])
     end
   end
 
@@ -109,17 +109,17 @@ describe Chef::Provider::User::Linux do
     end
 
     it "by default inactive is nil and we use ''" do
-      expect( provider.useradd_options ).to eql([""])
+      expect( provider.universal_options ).to eq([])
     end
 
     it "setting inactive to nil includes ''" do
       @new_resource.inactive nil
-      expect( provider.useradd_options ).to eql([""])
+      expect( provider.universal_options ).to eq([])
     end
 
     it "setting inactive to 90 includes -f" do
       @new_resource.inactive 90
-      expect( provider.useradd_options ).to eql(["-f"])
+      expect( provider.universal_options ).to eql(["-f", 90])
     end
   end
 end
```

Stromweld replied:
Thanks, that worked and all unit tests are passing now.

Stromweld commented:
Unit tests are added and passing now. Current failures look to come from latest merge from master and are unrelated to this code.

Stromweld commented on lib/chef/provider/user.rb:
```diff
@@ -66,13 +66,25 @@ def load_current_resource
+                              Time.at(shadow_info.sp_expire * 60 * 60 * 24).strftime("%Y-%m-%d")
+                            end
+              current_resource.expire_date(expire_date)
+            end
```
broken out as asked for, thanks.

jaymzh reviewed (approved):
LGTM. Thanks!

PrajaktaPurohit commented:
Looks like there are some failing tests in the verify pipeline that seem related to the changes in this PR: https://buildkite.com/chef-oss/chef-chef-main-verify/builds/12384#ee9d34d9-c443-41be-9f14-9095e5819abc/335-9004

jaymzh commented:
God love tests. They caught a pretty big bug we all missed.

You reference `current_resource.expire_date(expire_date)` in `provider/user.rb`, the generic user provider, but you only actually define it in the linux-specific resource, `linux_user.rb`, so when user gets used on non-linux platforms, that code crashes.

There's a few options:

1. Move `expire_date` into the generic provider, ignore it on non-supported platforms and throw a warning if it's used on one. If you do this, you should at least attempt to make `load_shadow_options` a bit more generic - at least all Unix-like OSes. Doesn't really make sense on Mac/Windows. If you do this, you should move the call to `load_shadow_options` to the generic provider, and drop your `load_current_resource` override.
2. Since you *only* ever call `load_shadow_options from the Linux provider, it makes sense to only define it in the Linux provider. This is the easiest option.

jaymzh reviewed (changes requested):
Yay for tests catching bugs.

Stromweld commented:
@jaymzh Thinking back on it, after talking to lamont a while back I believe he recommended putting the load_shadow_options in the generic user for future use with other OSes besides linux. Since linux is the only one calling it, it doesn't affect others. I did try moving it to linux provider and still got the same linux functional errors for user in group spec in build kite.

Stromweld commented:
In this last merge I tried testing moving the new properties from resource/user/linux_user.rb to resource/user.rb and that fixed the issues. The only 2 macos failures appear to be a network issue on the 1 job and a git package issue on the other.

@PrajaktaPurohit and @jaymzh, So question now is do we leave it this way or figure out why they aren't loading from linux_user.rb?

jaymzh commented:
Right, so that was `option 1` - move expire_date to the generic resource (I said provider, sorry, meant resource).

To be clear here, it was failing in `inactive` call on like 86 of user.rb.

As I said earlier, if we want to move the properties to the generic resource, then we should probably *also* warn if they are ever used on non-linux platforms (and, ideally, try to see if this will work on bsd and friends).

But that leaves your (very good) question: why the hell, on Ubuntu, did we get the generic resource instead of the OS-specific resource? And I suspect it's because in the test, we're calling `user` very specifically:
```ruby
              user(test_user).run_action(:create)
```

What's `user`, you ask?

```ruby
  def user(user)
    node = Chef::Node.new
    node.consume_external_attrs(ohai.data, {})
    run_context = Chef::RunContext.new(node, {}, Chef::EventDispatch::Dispatcher.new)
    usr = Chef::Resource.resource_for_node(:user, node).new(user, run_context)
    usr.password("ComplexPass11!") if windows?
    usr
  end
```

Which calls `resource_for_node`, which might-could-should do the right thing, except the `node` we make there doesn't seem to have the relevant data. Which is odd, because we do `node.consume_external_attrs(ohai.data, {})` which seems to have relevant data in it. That's very strange.

@PrajaktaPurohit - we should figure out what's up with that test - it could be a bug in the test or in the way we lazy-resolve the resource.

Still, if we're going to have `load_shadow_options` in the generic provider, which makes sense as lots of platforms have shadow info, then the properties it expects should be in the generic resource, so I think this change is fine. Or, to put it a different way, they both should either be in the generic code, or both should be in the Linux-specific code. In an effort to prevent someone from making a *different* implementation in a specific-provider later, lets go ahead and keep them in the generic provider as you've done.

But if we're going to do that, as I said before, lets make sure that when some poor shmuck sets `expire_date` in Solaris, it doesn't silently fail and cause a security issue. Lets make sure there's warnings there. Probably in `define_resource_requirements`.

Stromweld commented on lib/chef/provider/user.rb:
```diff
@@ -95,6 +107,11 @@ def define_resource_requirements
           a.whyrun "ruby-shadow is not installed. Attempts to set user password will cause failure.  Assuming that this gem will have been previously installed." \
             "Note that user update converge may report false-positive on the basis of mismatched password. "
         end
+        requirements.assert(:all_actions) do |a|
+          a.assertion { !((new_resource.expire_date || new_resource.inactive) && platform_family?(%w{aix freebsd mac_os_x netbsd openbsd openindiana solaris2 windows})) }
+          a.failure_message Chef::Exceptions::User, "Properties expire_date and inactive are not supported by this OS or have not been implemented for this OS yet."
+          a.whyrun "Properties expire_date and inactive are ignored as they are not supported by this OS or have not been implemented yet for this OS"
+        end
```
@jaymzh is this what you were looking for for a warning?

jaymzh commented on lib/chef/provider/user.rb:
```diff
@@ -95,6 +107,11 @@ def define_resource_requirements
             "Note that user update converge may report false-positive on the basis of mismatched password. "
         end
+        requirements.assert(:all_actions) do |a|
+          a.assertion { !((new_resource.expire_date || new_resource.inactive) && platform_family?(%w{aix freebsd mac_os_x netbsd openbsd openindiana solaris2 windows})) }
```
don't blacklist, that's silly. Do a.assertion { ... || platform_family?('linux') }` - and I think there's actually `linux?` helper you can just do instead.

Stromweld replied:
how does this look now?

jaymzh reviewed (changes requested):
Yup! Minor logic change, but yeah

jaymzh commented on lib/chef/provider/user.rb:
```diff
@@ -108,7 +108,7 @@ def define_resource_requirements
         end
         requirements.assert(:all_actions) do |a|
-          a.assertion { !((new_resource.expire_date || new_resource.inactive) && platform_family?(%w{aix freebsd mac_os_x netbsd openbsd openindiana solaris2 windows})) }
+          a.assertion { !((new_resource.expire_date || new_resource.inactive) && linux?) }
```
That's the opposite of what you want/ You want
```ruby
          a.assertion { !((new_resource.expire_date || new_resource.inactive) && !linux?) }
```

Which is a weird way of writing:

```ruby
          # either neither linux-only value is set, or we need to be on Linux.
          a.assertion { (!new_resource.expire_date && !new_resource.inactive) || !linux?) }
```

jaymzh replied:
Hmmmm - wonder if we can wrap a test around this...

Stromweld replied:
good catch, thanks. change made.

jaymzh reviewed (approved):
LGTM
